servers/services/src/main/java/org/projectnessie/services/impl/ContentApiImpl.java

```diff
@@ -54,19 +57,43 @@ public ContentApiImpl(
 
   @Override
   public ContentResponse getContent(
-      ContentKey key, String namedRef, String hashOnRef, boolean withDocumentation)
+      ContentKey key,
+      String namedRef,
+      String hashOnRef,
+      boolean withDocumentation,
+      boolean forWrite)
       throws NessieNotFoundException {
     try {
       ResolvedHash ref =
           getHashResolver()
               .resolveHashOnRef(namedRef, hashOnRef, new HashValidator("Expected hash"));
-      ContentResult obj = getStore().getValue(ref.getHash(), key, false);
+      ContentResult obj = getStore().getValue(ref.getHash(), key, forWrite);
       BatchAccessChecker accessCheck = startAccessCheck();
-      if (obj != null) {
-        accessCheck.canReadEntityValue(ref.getValue(), obj.identifiedKey()).checkAndThrow();
+
+      NamedRef r = ref.getValue();
+      accessCheck.canViewReference(r);
+      if (forWrite) {
+        accessCheck.canCommitChangeAgainstReference(r);
+      }
+
+      if (obj != null && obj.content() != null) {
+        accessCheck.canReadEntityValue(r, obj.identifiedKey());
+        if (forWrite) {
+          accessCheck.canUpdateEntity(r, obj.identifiedKey());
+        }
+
+        accessCheck.checkAndThrow();
+
         return ContentResponse.of(obj.content(), makeReference(ref), null);
       }
-      accessCheck.canViewReference(ref.getValue()).checkAndThrow();
+
+      if (forWrite) {
+        accessCheck
```

Comment:
Isn't this check redundant with the checks line 80-82?

Reply:
Not in every case (when the object does not exist and forWrite == false)